feat: add explanation for ignore directives

chengr4 opened the pull request:
Hi,

This is my first contribute in deno codebase. I feel nervous and hope it can go well.

This PR is for #1076.

Please check my updates and slowly guide me to the right direction 🙏. Thanks in advance

## TODO

- [x] Add test case for `test_parse_global_ignore_directives` (done at 2024.01.20)
- [ ] Feature: add multi-line

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/denoland/deno_lint?pullRequest=1229) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/denoland/deno_lint?pullRequest=1229) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/denoland/deno_lint?pullRequest=1229) it.</sub>

bartlomieju commented:
Looks like a good start. Please run `tools/format.ts` to format the code and make CI happy.

chengr4 commented:
hi @bartlomieju,
I fixed CI and added test cases for `test_parse_global_ignore_directives`. Please check it out when you get time, and tell me anything I missed 🙏.

BTW, There are two things I have zero confidence.

1. Variable Naming: eg. `comment_text_without_reason` and  `IGNORE_COMMENT_REASON_RE` <= I hope this is ok.
2. 
```rust
let comment_text =
        IGNORE_COMMENT_CODE_RE.replace_all(&comment_text_without_reason, ",");
```
About adding `&` in front of `comment_text_without_reason`, I only followed the instruction of the compiler and hope it was the best practice.

bartlomieju commented on src/ignore_directives.rs:
```diff
@@ -197,14 +205,17 @@ export function deepAssign(
-object | undefined {}
+Object | undefined {}
+
+// deno-lint-ignore no-explicit-any no-empty no-debugger -- reason for ignoring
+function foo(): any {}
```
This looks good, I have only one worry - it will be easy to make comments over 80 or 100 characters with explanation. I wonder if we could figure out a way to make these comment multi-line.

Could you try to add a test case for something like:
```
// deno-lint-ignore no-explicit-any no-empty no-debugger -- super, duper, very long
// reason for ignoring
function foo(): any {}
```

chengr4 replied:
I tried the test case you gave. it was able to be passed.
but `dlint` with real TS file did not work.

Do you want to finish multi-line issue in this PR? Or we can open a new issue for it with new PR.

P.s. A small question: Is it common to add comments under 'lint ignore' (Is it common in Node.js)? I personally have never seen it before.

bartlomieju replied:
Yeah, we can do multi-line support in a follow up. Let's land it as is.

bartlomieju commented:
> hi @bartlomieju, I fixed CI and added test cases for `test_parse_global_ignore_directives`. Please check it out when you get time, and tell me anything I missed 🙏.
> 
> BTW, There are two things I have zero confidence.
> 
> 1. Variable Naming: eg. `comment_text_without_reason` and  `IGNORE_COMMENT_REASON_RE` <= I hope these are ok.
> 
> ```rust
> let comment_text =
>         IGNORE_COMMENT_CODE_RE.replace_all(&comment_text_without_reason, ",");
> ```
> 
> About adding `&` in front of `comment_text_without_reason`, I only followed the instruction of the compiler and hope it was the best practice.

Yeah, these seems fine 👍

chengr4 commented:
Sorry, I pressed the button accidentally. I didn't intend to rush you 😓.

bartlomieju reviewed (approved):
LGTM, nice work @chengr4!
